x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_upgrade/route.ts

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { pickBy } from 'lodash';
+import { transformError } from '@kbn/securitysolution-es-utils';
+
+import type { PrebuiltRuleAsset } from '../../model/rule_assets/prebuilt_rule_asset';
+import { REVIEW_RULE_UPGRADE_URL } from '../../../../../../common/detection_engine/prebuilt_rules';
+import type {
+  ReviewRuleUpgradeResponseBody,
+  RuleUpgradeInfoForReview,
+  RuleUpgradeStatsForReview,
+} from '../../../../../../common/detection_engine/prebuilt_rules/api/review_rule_upgrade/response_schema';
+import type { PrebuiltRuleVersionInfo } from '../../model/rule_versions/prebuilt_rule_version_info';
+import type {
+  CalculateRuleDiffArgs,
+  CalculateRuleDiffResult,
+} from '../../logic/diff/calculate_rule_diff';
+import { calculateRuleDiff } from '../../logic/diff/calculate_rule_diff';
+import type { ThreeWayDiff } from '../../../../../../common/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff';
+import type { RuleResponse } from '../../../../../../common/detection_engine/rule_schema';
+
+import type { SecuritySolutionPluginRouter } from '../../../../../types';
+import { buildSiemResponse } from '../../../routes/utils';
+
+import { createPrebuiltRuleAssetsClient } from '../../logic/rule_assets/prebuilt_rule_assets_client';
+import { createPrebuiltRuleObjectsClient } from '../../logic/rule_objects/prebuilt_rule_objects_client';
+import { getVersionBuckets } from '../../model/rule_versions/get_version_buckets';
+
+export const reviewRuleUpgradeRoute = (router: SecuritySolutionPluginRouter) => {
+  router.post(
+    {
+      path: REVIEW_RULE_UPGRADE_URL,
+      validate: {},
```

Comment:
Discussed offline. @xcrzx is going to create a separate task for this optimization. There's a doubt that it's achievable, and if yes, reasonable in terms of performance/efficiency. Still, it would make sense to research this option.